src/raster/warp/warp_options.rs

```diff
@@ -0,0 +1,300 @@
+use std::fmt::{Debug, Display, Formatter};
+use std::mem::transmute;
+use std::ptr::NonNull;
+
+use crate::cpl::CslStringList;
+use crate::errors::{GdalError, Result};
+use crate::raster::warp::resample::WarpResampleAlg;
+use crate::raster::GdalDataType;
+use crate::utils::_last_null_pointer_err;
+use crate::xml::GdalXmlNode;
+use gdal_sys::{
+    GDALCloneWarpOptions, GDALCreateWarpOptions, GDALDeserializeWarpOptions,
+    GDALDestroyWarpOptions, GDALSerializeWarpOptions, GDALWarpInitDefaultBandMapping,
+    GDALWarpInitDstNoDataReal, GDALWarpInitSrcNoDataReal, GDALWarpOptions,
+    GDALWarpResolveWorkingDataType,
+};
+use libc::c_char;
+
+/// Container for options provided to GDAL Warp routines.
+///
+/// See: [`GDALWarpOptions`](https://gdal.org/api/gdalwarp_cpp.html#_CPPv415GDALWarpOptions)
+/// for details.
+pub struct GdalWarpOptions(NonNull<GDALWarpOptions>);
+
+impl GdalWarpOptions {
+    pub fn new() -> Self {
+        unsafe { Self::from_ptr(GDALCreateWarpOptions()) }
+    }
+
+    /// Create Self from a raw pointer.
+    ///
+    /// # Safety
+    /// Caller is responsible for ensuring `ptr` is not null, and
+    /// ownership of `ptr` is properly transferred
+    pub unsafe fn from_ptr(ptr: *mut GDALWarpOptions) -> Self {
+        Self(NonNull::new_unchecked(ptr))
+    }
+
+    /// Specify the resampling algorithm to use in Warp operation.
+    pub fn with_resampling_alg(&mut self, alg: WarpResampleAlg) -> &mut Self {
+        unsafe { (*self.as_ptr_mut()).eResampleAlg = alg.to_gdal() };
+        self
+    }
+
+    /// Get the resampling algorithm to be used in Warp operation.
+    pub fn resampling_alg(&self) -> WarpResampleAlg {
+        // `unwrap` below is ok because `with_resampling_alg` is the only way it got set,
+        // aside from the GDAL default, which is `GRA_NearestNeighbour`.
+        WarpResampleAlg::from_gdal(unsafe { (*self.as_ptr()).eResampleAlg }).unwrap_or_default()
+    }
+
+    /// Set the datatype used during processing.
+    ///
+    /// If unset, the algorithm picks the datatype.
+    pub fn with_working_datatype(&mut self, dt: GdalDataType) -> &mut Self {
+        unsafe { (*self.as_ptr_mut()).eWorkingDataType = dt.gdal_ordinal() };
+        self
+    }
+
+    /// Fetch working datatype, if specified.
+    pub fn working_datatype(&self) -> Option<GdalDataType> {
+        let c_dt = unsafe { (*self.as_ptr()).eWorkingDataType };
+        let dt: GdalDataType = c_dt.try_into().ok()?;
+
+        // Default is `Unknown`, so we consider that "unspecified".
+        if dt == GdalDataType::Unknown {
```

Comment:
This might be a little confusing, should we drop the `Option`?

Reply:
If we did, would we change this

```rust
let dt: GdalDataType = c_dt.try_into().ok()?;
```

to 

```rust
let dt: GdalDataType = c_dt.try_into().unwrap_or(GdalDataType::Unknown);
```
?